File: contest7/pE.py
def dfs(adj, visited, cats, s, tot, m, count):
    if visited[s]:
        return
    visited[s] = True
    if cats[s]:
        tot += 1
    else:
        tot = 0
    if tot > m:
        return
    if s != 1 and len(adj[s]) == 1:
        count[0] += 1
    for i in adj[s]:
        dfs(adj, visited, cats, i, tot, m, count)
    return


n, m = map(int, input().split())
adj = [[] for i in range(n + 1)]
visited = [False] * (n + 1)
cats = [0] + list(map(int, input().split()))
for i in range(n - 1):
    x, y = map(int, input().split())
    adj[x].append(y)
    adj[y].append(x)
count = [0]
dfs(adj, visited, cats, 1, 0, m, count)
print(count[0])

File: contest7/pE.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(vector<vector<int>>& adj, vector<bool>& seen, vector<int>& cats, int s, int tot, int m, int& leaves) {
    if (seen[s])
        return;
    seen[s] = true;
    if (cats[s])
        tot += 1;
    else
        tot = 0;
    if (tot > m)
        return;
    if (s != 1 && adj[s].size() == 1)
        leaves += 1;
    for (int i : adj[s])
        dfs(adj, seen, cats, i, tot, m, leaves);
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    // freopen("input.txt", "r", stdin);
    /// bibliografia: guide to competitive programming
    /*
        revisar:
        getline() ///hay distintos parametros aqui.

    */
    int n, m;
    cin >> n >> m;

    vector<vector<int>> adj(n + 1);
    vector<bool> seen(n + 1, false);
    vector<int> cats(n + 1);

    for (int i = 1; i <= n; ++i)
        cin >> cats[i];

    for (int i = 0; i < n - 1; ++i) {
        int x, y;
        cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }

    int leaves = 0;
    dfs(adj, seen, cats, 1, 0, m, leaves);

    cout << leaves << "\n";

    return 0;
}
